core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -1833,6 +1832,9 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean, dynamicConfigO
       s"Found ${advertisedListenerNames.map(_.value).mkString(",")}. The valid options based on the current configuration " +
       s"are ${listenerNames.map(_.value).mkString(",")}"
     )
+    // Ensure controller listeners are not in the advertised listeners list
+    require(!controllerListeners.exists(advertisedListeners.contains),
```

Comment:
Hmm, not sure about this. In general, listener specifies the host to bind and advertised listener specifies the host for the client to connect. These two hosts can be different and the same may apply to the controller listener.

For example, in CCloud, currently we have the following where the binding host is different from the connecting host.

advertised.listeners = INTERNAL://kafka-34.kafka.pkc-4n3dv.svc.cluster.local:9071,REPLICATION://kafka-34.kafka.pkc-4n3dv.svc.cluster.local:9072,EXTERNAL://b34-pkc-4n3dv.us-east-1.aws.confluent.cloud:9092

listeners = INTERNAL://:9071,REPLICATION://:9072,EXTERNAL://:9092

Reply:
The issue here is that "advertised listeners" is used in a lot of places to mean "listeners that clients for the broker should connect to."  And we definitely don't want the client trying to connect to controller listeners.  We could change how advertised listeners is used, but I think a lot of places in the code would have to change.  How about introducing a config like "advertised controller listeners" instead?

Reply:
Yes, for the controller listener, only other brokers need to connect to it. The question is whether we want the flexibility for the binding host and the connecting host to be different (e.g. binding to all interface and connecting to the private IP). If we do, we need the controller listener to be defined in both listeners and advertised.listeners. To avoid exposing the controller listener to the clients, it seems that we can just exclude it from the metadata response?

Reply:
The kip-500 controller doesn't really advertise its listeners, though.  It doesn't appear in metadata responses.  So you should be able to get the result you want by specifying the listener as 0.0.0.0://9093 on the controller itself, and example.com://9093 on the brokers that are connecting to the controller.

Reply:
@cmccabe : Good point. Since the advertised host is specified in controlller.connect, we don't need it in the advertised listeners. So, this change is fine to me.

Reply:
@cmccabe Do you think this is good to merge now?